File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
#닉네임 중복 허용, 닉네임 변경 가능, 변경하면 이전에 나갔던 기록까지 불러와 닉네임 변경해야 함

#uid는 dict의 key, 닉네임은 value로 저장
def solution(record):
    user = {}
    methods = {"Enter":"님이 들어왔습니다.", "Leave":"님이 나갔습니다."}
    result = []
    for r in record:
        mun = r.split(" ") #[method, uid, nickname]
        if mun[0] != "Leave":
            user[mun[1]] = mun[2]
    for r in record:
        mun = r.split(" ") #[method, uid, nickname]
        if mun[0] in methods:
            result.append('{}{}'.format(user[mun[1]],methods[mun[0]]))
    return result

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        Map<String,String> m = new HashMap<>();
        for(int i=0;i<record.length;i++){
            String[] tmp = record[i].split(" ");
            if(!tmp[0].equals("Leave")) {
                m.put(tmp[1],tmp[2]);
            }
        }
        List<String> rtn = new ArrayList<>();
        for(int i=0;i<record.length;i++){
            String[] tmp = record[i].split(" ");
            switch(tmp[0]){
                    case "Enter"->rtn.add(m.get(tmp[1])+"님이 들어왔습니다.");
                    case "Leave"->rtn.add(m.get(tmp[1])+"님이 나갔습니다.");
            }
        }
        
        return rtn.toArray(new String[0]);
    }
}